platform/view/services/comm/comm_test.go

```diff
@@ -111,3 +141,76 @@ func testExchange(aliceNode, bobNode Node) {
 
 	wg.Wait()
 }
+
+func testExchangeManySenders(aliceNode, bobNode Node, numWorker, numOfMsgs int) {
+	var wgBob sync.WaitGroup
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// setup bob as our receiver
+	bobMasterSession, err := bobNode.commService.MasterSession()
+	Expect(err).ToNot(HaveOccurred())
+	wgBob.Add(1)
+	go func() {
+		defer wgBob.Done()
+		for {
+			select {
+			// run until we close via the context
+			case <-ctx.Done():
+				return
+			case response := <-bobMasterSession.Receive():
+				// get our message from master session
+				Expect(response).ToNot(BeNil())
+				Expect(response.Payload).To(Equal([]byte("ping")))
+
+				// create a response session
+				bobSession, err := bobNode.commService.NewSessionWithID(response.SessionID, "", response.FromEndpoint, response.FromPKID, nil, nil)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(bobSession.Send([]byte("pong"))).To(Succeed())
+
+				// close it
+				bobSession.Close()
+				Eventually(bobSession.Info().Closed).Should(BeTrue())
+			}
+		}
+	}()
+
+	// setup alice our sender
+	var wgAlice sync.WaitGroup
+	for i := 0; i <= numWorker; i++ {
+		wgAlice.Add(1)
+		go func() {
+			defer wgAlice.Done()
+
+			// we send every message in a fresh session
+			for j := 0; j <= numOfMsgs; j++ {
+				// setup
+				aliceSession, err := aliceNode.commService.NewSession("", "", rest.ConvertAddress(bobNode.address), bobNode.pkID)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(aliceSession.Info().Endpoint).To(Equal(rest.ConvertAddress(bobNode.address)))
+				Expect(aliceSession.Info().EndpointPKID).To(Equal(bobNode.pkID.Bytes()))
+
+				// send
+				Eventually(aliceSession.Send([]byte("ping"))).Should(Succeed())
+
+				// receive
+				Eventually(func(g Gomega) {
+					response := <-aliceSession.Receive()
+					g.Expect(response).ToNot(BeNil())
+					g.Expect(response).To(HaveField("Payload", Equal([]byte("pong"))))
+				}).Should(Succeed())
+
+				// close
+				aliceSession.Close()
```

Comment:
While writing this test, did you notice the logs were giving confusing messages? Or that we didn't shut down gracefully?

Maybe we should think of more test cases. For instance if we close the channel instead of sending back pong, what do we expect to receive on the other end?

Reply:
> While writing this test, did you notice the logs were giving confusing messages? Or that we didn't shut down gracefully?

Oh the test log is wilde and outputs many warnings. Example:
> 2025-04-11 13:03:03.957 CEST 9c66 WARN [view-sdk.services.comm.rest.host] readMessages -> Context for stream [5E1V9rQHbsYVziNe7bHPJktJjyQBQTjYYhucotHcmEr.127.0.0.1:57126.BnsK7rcN3qKRV8y7LYUyWE0jEbfTsauQ.] closed by us. Error: %!w(*errors.errorString=&{context canceled})